staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/list_width.go

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package request
+
+import (
+	"math"
+	"net/http"
+	"net/url"
+	"strconv"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+)
+
+const (
+	// the minimum number of seats a list request occupies
+	minimumSeatsList uint = 3
+
+	// the maximum number of seats a list request occupies
+	maximumSeatsList uint = 10
+)
+
+func newListWidthEstimator(countFn objectCountGetterFunc) WidthEstimatorFunc {
+	estimator := &listWidthEstimator{
+		countFn: countFn,
+	}
+	return estimator.estimate
+}
+
+type listWidthEstimator struct {
+	countFn objectCountGetterFunc
+}
+
+func (e *listWidthEstimator) estimate(r *http.Request) Width {
+	requestInfo, ok := apirequest.RequestInfoFrom(r.Context())
+	if !ok {
+		return Width{Seats: minimumSeatsList}
+	}
+
+	query := r.URL.Query()
+	count := e.countFn.Get(key(requestInfo))
+
+	var fieldSelectorPresent bool
+	if len(query.Get("fieldSelector")) > 0 {
+		fieldSelectorPresent = true
+	}
+
+	var resourceVersionNotSet bool
+	if version, ok := query["resourceVersion"]; ok && len(version) > 0 && len(version[0]) == 0 {
+		resourceVersionNotSet = true
+	}
```

Comment:
I'm assuming that with RV being present or not you want to cache the case whether the request is served from cache or not.
This logic is actually more complex than just RV set or not:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L603

To avoid risking divergence, what we should probably do is to move that logic to a helper method, e.g.:
```
staging/src/k8s.io/apiserver/pkg/util/chunking/chunking.go

func ListFromCache(opts *metav1.ListOptions) bool {
  // That logic goes here
}
```
and use it here.

Reply:
Actually - it's a bit more tricky, because that one is using storage.ListOptions...
So trying to share the code might become to cumbersome..

I opened https://github.com/kubernetes/kubernetes/pull/103537 to factor this logic into its own function.
Once that merges, let's just duplicat this logic here and add comments in both places saying that we should keep those in sync.